VChat: Redone chat output done in Vue.js

Arokha opened the pull request:
[It's a draft because it'd be best if we merged it downstream first and I could fix bugs live in the HTML and JS on the first round or two then give it to you with the Day-0 patches]

VChat is an overhauled chat output box in HTML with JS frameworks, similar to 'goonchat'. It stands for vorechat but you can pretend it stands for Vue.js chat if vore offends your naive sensibilities.

It features things like:
---
### Living chat history
I can't really make a GIF of this one but basically if you lose your connection or disconnect, your chat history is returned to you when you reconnect, **including things you would have seen were you connected**. So you don't miss any posts.

---
### Tabbed chat
![tabs](https://user-images.githubusercontent.com/15028025/74699472-c34df380-51ce-11ea-9933-d4f299f582f4.gif)
---
### Unread indicators
![image](https://user-images.githubusercontent.com/15028025/74699692-5be47380-51cf-11ea-90d7-b2beb7a3e1a0.png)
---
### Cute lil animations
![2020-02-17_21-50-21](https://user-images.githubusercontent.com/15028025/74699761-87fff480-51cf-11ea-8c42-0e33a83dee1f.gif)
---
### Dark mode
![darkmode](https://user-images.githubusercontent.com/15028025/74699482-cb0d9800-51ce-11ea-818d-dd541e4b2011.gif)
---
### Pausing scrollback
![pause](https://user-images.githubusercontent.com/15028025/74699493-d234a600-51ce-11ea-8707-8b6246f687ef.gif)
---
### Combining messages
![combining](https://user-images.githubusercontent.com/15028025/74699513-dcef3b00-51ce-11ea-87cd-821241bfba2e.gif)
---
### Some sizes
![sizes](https://user-images.githubusercontent.com/15028025/74699520-e37db280-51ce-11ea-81d4-a3b12aba59e0.gif)
---

Some things I'd like to add in the future:
* Saving the tabs between rounds
* Notification sounds, user settable?
* Ability to turn off notifications
* Other stuff???

zekesulastin commented:
Amazing o:

Is there anyone on your downstream who would be testing it in Linux?

Arokha commented:
> Is there anyone on your downstream who would be testing it in Linux?

Neerti tested it and it doesn't appear to work in WINE with IE8, it falls back to the old chat after 60s. I think it's pretty likely it requires ECMAScript 5 at least, which is IE9.

It might be possible to polyfill everything back to IE9 but... I don't know. It's *extremely difficult* to get the JS errors out of Byond's internal IE thing to see what's even going wrong.

Arokha commented:
I tried shimming ES5 for linux users but it turns out it's not something that's shimmable. It's the fact that IE8 lacks `Object.defineProperty`, and any framework) other than old versions of jquery-ui like goonchat uses) like Vue.js and React.js and AngularJS all only support IE9 or later.

So, Linux users need to figure out how to cram IE9 support in there.

Arokha commented:
Ok. It's working fine for us now after a couple weeks of tests and fixes.
Still no advice for Linux users. I feel for you. I wish I could make it work. Demand LummoxJR make a linux client.

Atermonera commented:
![good gods the diff](https://puu.sh/FkoOp/a8f84577c7.png)
Well, it's a good thing I've at least been exposed to js somewhere in the past 5 years...
(This will probably take more than a week for me to get through.)

Arokha commented:
Don't read the vue.js lol, it's just a library.

https://vuejs.org

Atermonera commented on code/modules/client/asset_cache.dm:
```diff
@@ -0,0 +1,287 @@
+	"}, "window=asset_cache_browser")
+
+	var/t = 0
+	var/timeout_time = (ASSET_CACHE_SEND_TIMEOUT * client.sending.len) + ASSET_CACHE_SEND_TIMEOUT
```
In the case where there's only one job being sent, it'll potentially wait up to `([timeout] * [1]) + [timeout] == 2 * [timeout]`

Arokha replied:
True, though it shouldn't hurt anything. Why not be extra nice to the newcomer? :p

Atermonera commented on code/modules/client/asset_cache.dm:
```diff
@@ -0,0 +1,287 @@
+//This proc blocks(sleeps) unless verify is set to false
+/proc/send_asset(var/client/client, var/asset_name, var/verify = TRUE)
+	if(!istype(client))
+		if(ismob(client))
```
Reviewing over multiple days so I may not remember correctly but didn't you add // isn't there a helper proc to extract the client from a mob?

Arokha replied:
A few different people wrote this, so they may not have all been aware of all the options available. Yes, I see a define `CLIENT_FROM_VAR(x)`.

Arokha commented on code/__defines/subsystems.dm:
```diff
@@ -63,13 +63,15 @@ var/global/list/runlevel_flags = list(RUNLEVEL_LOBBY, RUNLEVEL_SETUP, RUNLEVEL_G
 #define INIT_ORDER_CIRCUIT		-21
 #define INIT_ORDER_AI			-22
 #define INIT_ORDER_JOB			-23
+#define INIT_ORDER_CHAT			-100 //Should be last to ensure chat remains smooth during init.
```
No, the job of SSchat is just to enqueue messages for more graceful batching/sending of them. Until SSchat is initialized, it doesn't bother doing that. If you look in `vchat_client.dm` you'll see it in __to_chat:

```
	// Now lets either queue it for sending, or send it right now
	if(Master.current_runlevel == RUNLEVEL_INIT || !SSchat?.subsystem_initialized)
		to_chat_immediate(target, world.time, message)
	else
		SSchat.queue(target, world.time, message)
```

Arokha commented on code/controllers/subsystems/chat.dm:
```diff
@@ -0,0 +1,87 @@
+		var/list/messages = msg_queue[C]
+		msg_queue -= C
+		if (C)
+			C << output(jsEncode(messages), "htmloutput:putmessage")
```
Not sure how chat messages are 'handled' with the stream operator (aka `client << "message"`), since it's an internal byond function. If the MC is totally destroyed but the server is not frozen, then chat messages would stop with the advent of SSchat. In SSchat-world or pre-SSchat, messages would stop if the server was deadlocked either way.

Movement would continue (until someone ports SSinput anyway). But the game would be unplayable: machinery would no longer function, mobs would not get life() ticks, etc. I would not worry too much about chat messages in a world where the MC is not running, because that would be a short-lived round anyway.

This did happen once downstream, before we realized the bug with MC and services that have both `SS_BACKGROUND` and `SS_NO_TICK_CHECK` crashing the MC (which is the case on the inactivity subsystem until we changed it). People wrote things on paper and passed them back and forth. It was funny, though the round couldn't really continue because the MC was gone and things like doors and stuff no longer functioned since they are machinery. Restarting the MC fixes the issue and sends all enqueued messages.

Arokha commented on code/controllers/subsystems/chat.dm:
```diff
@@ -0,0 +1,87 @@
+			return
+
+/datum/controller/subsystem/chat/stat_entry()
+	..("C:[msg_queue.len]")
```
It's unlikely to be high, or probably even > 0, if things are working fine. Depends on when `/mob/Stat()` is called vs when SSchat runs. I'm not 100% sure actually on how those are timed.

Arokha commented on maps/RandomZLevels/wildwest.dm:
```diff
@@ -1,173 +1,173 @@
+
+	if(istype(M, /mob/living/carbon/human) || istype(M, /mob/living/carbon/monkey))
+		for(var/mob/O in viewers(world.view, src.loc))
+			to_chat(O, "<font color='red'>[M] triggered the [bicon(src)] [src]</font>")
```
lol

Arokha commented on interface/skin.dmf:
```diff
@@ -2,735 +2,553 @@ macro "borghotkeymode"
 	elem 
 		name = "TAB"
 		command = ".winset \"mainwindow.macro=borgmacro hotkey_toggle.is-checked=false input.focus=true input.background-color=#D3B5B5\""
-		is-disabled = false
```
Nothing, because is-disabled being false is the default.

I didn't actually remove those, Byond did. If you just open skin.dmf in DreamMaker right now and hit Ctrl-S, it will remove all those because it knows they aren't necessary.

I assume they were put there by someone editing skin.dmf in an older version of DreamMaker that didn't know any better and it put them in.

Arokha commented:
I think your Travis is broke. It seems like your `master` doesn't compile, cus I didn't even change that file there.

```
loading interface/skin.dmf
code/modules/mob/living/silicon/robot/robot.dm:908:error: =: conflicts with previous definition
polaris.dmb - 1 error, 0 warnings (3/23/20 2:07 am)
```

Atermonera commented:
What's most fascinating about that error is that it's always _exactly_ that error that crops up when travis falls over. Which may say something about the process of compiling the code, but I'm not so clever as to be able to decipher such from "Whenever travis needs to be kicked, it's always for X". Generally whenever I scan github, if I spot a PR with a failed build or check, I'll take a look and see if travis needs to be kicked, especially if it's marked as unread

Atermonera commented:
I have no idea what's up with those conflicts, but another question:
If one were to use a << for, say, printing statements to world while testing code, would that break horribly, would it just print to all tabs, or what?

Arokha commented:
>If one were to use a << for, say, printing statements to world while testing code, would that break horribly, would it just print to all tabs, or what?

It would not do anything. You'd want to use to_world, or world.log instead.

Arokha commented:
Convince new Travis to build my PR

Novacat commented on code/modules/client/asset_cache.dm:
```diff
@@ -0,0 +1,271 @@
+
+	var/list/common_dirs = list(
+		"nano/css/",
+		"nano/images/",
```
```suggestion
		"nano/images/",
		"nano/images/modular_computers/",
```

Atermonera commented:
Completed a test round, requests, reports of bugs and wonky things as follows, in rough order of them being reported:
(Using links instead of embedded images because this PR's conversation is already pretty long, don't need to make it 5 miles with screenshots of broken examine text)
[Linux Chat icons broken](https://i.imgur.com/C5nZ0Db.png)

[AdminPMs sorted poorly](https://puu.sh/FoUWO/d737f23fc1.png)
> The initial message goes to Notices,
> Any responses don't go to Admin PMs

Separate config for darkmode colors?
> AdminPM responses are too dark

Verb/option that switches to oldchat?

Command reports / centcom messages not showing on Main tab
> Captain's office computer doesn't work
> Admin Announce does work

[Text/Info/Wiki/Forum/etc buttons at top look weird in dark mode w/o margin](https://puu.sh/FoV8d/d653d37b4e.png)

[Oldchat looc color is now black?](https://i.imgur.com/1ecDt1o.png)

> [2020-03-25T22:05:21] Runtime in vchat_client.dm,58: bad client
>   proc name: update vis (/datum/chatOutput/proc/update_vis)
>   src: /datum/chatOutput (/datum/chatOutput)
>   call stack:
>   /datum/chatOutput (/datum/chatOutput): update vis()
>   /datum/chatOutput (/datum/chatOutput): become broken()
>   /datum/chatOutput (/datum/chatOutput): load()

Examine tab doesn't like images
>  Shows the <img> tag for some, unreliable, and no image for others
> [Link 1](https://puu.sh/FoVBZ/56480c42ed.png)
> [Link 2](https://puu.sh/FoVCm/d663201777.png)
> [Link 3](https://puu.sh/FoWYI/a95ad155ef.png)

[NTNet/baycomputer icons are broken](https://cdn.discordapp.com/attachments/182550163263324160/692498271888736276/unknown.png) Apparently already fixed?

.ping doesn't go to any tabs, not even Main

Verb outputs don't go to any particular tabs
>  Advanced who
>  Check round info

[Radio icons don't show well against dark background](https://puu.sh/FoWCM/e62cc74bbc.png)

Opening a large number of tabs and then closing out of most of them kicks the user

Arokha commented:
You might find some of my answers are 'Goonchat hasn't done this either' to try and show that it's not just my ineptitude that's at fault, and it's more likely something that's not possible.

> Linux Chat icons broken

Setting `enable-http-images` on the OUTPUT pane ... maaaaay fix this. I'm not 100% sure. It may also open up vulnerabilities to weird image injection and getting people's IPs from other clients via shenanigans? I'm not sure. I'll have to think about it.

> AdminPMs sorted poorly

Fixable by applying the proper <span> classes in code (this is basically my answer for anything 'goes to the wrong filter category'. It's reliant on people using the proper classes on span tags everywhere else in the code, as vchat has to guess based on those. In this case, maybe it's `pm`)

> Separate config for darkmode colors?

Already a thing, look in ss13styles.css, for example:
```css
.deadsay				{color: #530FAD;}
.inverted .deadsay		{color: #732FCD;} /* Dark mode */
```
Deadchat gets a different color in dark mode. Just apply whatever separate darkmode colors you'd like to the spans. You may find that a lot of bad-looking places in the code (including those admin replies), someone has decided to use <font color='blue'> instead of a span class. Those would need to be replaced with an appropriate class. (this will also fix sorting).

> Verb/option that switches to oldchat?
Mnnnnnn. I'm opposed to maintaining two systems and trying to make them both work, especially when one is really meant as a fallback. For example, Goonchat has a verb to 'switch' kinda, but it's actually a 'my goonchat is broken' button and tries to walk you through several steps of fixing each time it before letting you use the old chat. I get the feeling they feel the same as I do about it. You can add one if you want, but I won't be trying to maintain two different chats at the same time.

> Command reports / centcom messages not showing on Main tab
Oversight, as we don't use your style of command reports. We use radio transmissions as command reports instead, Eris style. I'll make a commit to fix it for you.

> Text/Info/Wiki/Forum/etc buttons at top look weird in dark mode w/o margin
They look fine to me... but you're welcome to adjust skin.dmf to alter their appearance? I don't really want to get into a back-and-forth pretending I'm a graphic designer about what pixels are better where so I'd suggest you just make that change, no offense. It would be a lot easier than trying to 'explain' what you think looks better.

> Oldchat looc color is now black?
Likely related to the bizarre class choices they had and what I changed them to for LOOC. Probably fixable, I think.

> [2020-03-25T22:05:21] Runtime in vchat_client.dm,58: bad client
When you disconnect before vchat finishes loading, it does this. Doesn't hurt anything since the runtime is on a doomed chatOutput object about to be deleted, but I'll add a check to remove the runtime.

> Examine tab doesn't like images

That's only an issue on the bar sign as far as I'm aware. And walls don't have real icons. They're a composite of overlays. You'll notice that it's not correct prior to vchat either and just shows an arbitrary floor/plating icon:
![image](https://user-images.githubusercontent.com/15028025/77597232-94652480-6ed4-11ea-89e6-613c7140c69d.png)

Between 'arbitrary icon' and 'no icon' I don't think there's much difference tbh.

> NTNet/baycomputer icons are broken Apparently already fixed?

Yes, you started your test before Ascian pointed out I missed a fix commit from VS. I added it.

> .ping doesn't go to any tabs, not even Main

Unfixable in any HTML chat. Goonchat in every server also has this. You can click the little colored dot in the corner for a quick latency check, though it's not precisely network latency. It's the time it takes for the server to reply to some javascript code via Topic.

> Verb outputs don't go to any particular tabs

Put <span> tags around their outputs! I guesssssss I can do it for you.

> Radio icons don't show well against dark background

... okay ... I don't know how that can be helped really. The icons are tiny. You can increase font and line size to allow the icons to be larger, if you want?

> Opening a large number of tabs and then closing out of most of them kicks the user

Hmmm. Closing tabs shouldn't call Topic(). I assume you mean it kicks them with the 'stop spamming Topic()' message? Or no?

Atermonera commented:
The wall examine tab was particularly notable because it was an r-wall, and the "tool-icon to continue deconstructing" was just [blank] to deconstruct. And the kick message was something on the order of "Too many messages to/from the server"

Atermonera commented:
Because I just thought of it, and since I personally can't catch it in live tests because devs don't get admin logs (and I'm our only just-a-dev), can admin logs, debug logs, etc, be filtered, and if so what filters do they go to?

And semi-related, what's actually involved in adding new filters?

Atermonera commented on code/modules/examine/examine.dm:
```diff
@@ -34,7 +34,7 @@
 // Quickly adds the boilerplate code to add an image and padding for the image.
 /proc/desc_panel_image(var/icon_state)
-	return "\icon[description_icons[icon_state]][EXAMINE_PANEL_PADDING]"
+	return "[bicon(description_icons[icon_state])][EXAMINE_PANEL_PADDING]"
```
This is specifically the change that seems to have broken the examine tab, at least wrt wall deconstruction icons

Arokha commented:
> Because I just thought of it, and since I personally can't catch it in live tests because devs don't get admin logs (and I'm our only just-a-dev), can admin logs, debug logs, etc, be filtered, and if so what filters do they go to?

> And semi-related, what's actually involved in adding new filters?

Don't know actually. Filtering is done based on the first <span> tag in the message vchat gets from Byond, and the classes it has on it. If you look in vchat.js, there's an array of the categories:
```js
//The table to map game css classes to our vchat categories
type_table: [
	{
		matches: ".say, .emote",
		becomes: "vc_localchat",
		pretty: "Local Chat",
		tooltip: "In-character local messages (say, emote, etc)",
		required: false,
		admin: false
	},
	{
		matches: ".alert, .syndradio, .centradio, .airadio, .entradio, .comradio, .secradio, .engradio, .medradio, .sciradio, .supradio, .srvradio, .expradio, .radio, .deptradio, .newscaster",
		becomes: "vc_radio",
		pretty: "Radio Comms",
		tooltip: "All departments of radio messages",
		required: false,
		admin: false
	},
	{
		matches: ".notice, .adminnotice, .info, .sinister, .cult",
		becomes: "vc_info",
		pretty: "Notices",
		tooltip: "Non-urgent messages from the game and items",
		required: false,
		admin: false
	},
	{
		matches: ".critical, .danger, .userdanger, .warning, .italics",
		becomes: "vc_warnings",
		pretty: "Warnings",
		tooltip: "Urgent messages from the game and items",
		required: false,
		admin: false
	},
	{
		matches: ".deadsay",
		becomes: "vc_deadchat",
		pretty: "Deadchat",
		tooltip: "All of deadchat",
		required: false,
		admin: false
	},
	{
		matches: ".ooc:not(.looc)",
		becomes: "vc_globalooc",
		pretty: "Global OOC",
		tooltip: "The bluewall of global OOC messages",
		required: false,
		admin: false
	},
	//VOREStation Add Start
	{
		matches: ".nif",
		becomes: "vc_nif",
		pretty: "NIF Messages",
		tooltip: "Messages from the NIF itself and people inside",
		required: false,
		admin: false
	},
	//VOREStation Add End
	{
		matches: ".pm",
		becomes: "vc_adminpm",
		pretty: "Admin PMs",
		tooltip: "Messages to/from admins ('adminhelps')",
		required: false,
		admin: false
	},
	{
		matches: ".admin_channel",
		becomes: "vc_adminchat",
		pretty: "Admin Chat",
		tooltip: "ASAY messages",
		required: false,
		admin: true
	},
	{
		matches: ".mod_channel",
		becomes: "vc_modchat",
		pretty: "Mod Chat",
		tooltip: "MSAY messages",
		required: false,
		admin: true
	},
	{
		matches: ".event_channel",
		becomes: "vc_eventchat",
		pretty: "Event Chat",
		tooltip: "ESAY messages",
		required: false,
		admin: true
	},
	{
		matches: ".ooc.looc, .ooc .looc", //Dumb game
		becomes: "vc_looc",
		pretty: "Local OOC",
		tooltip: "Local OOC messages, always enabled",
		required: true
	},
	{
		matches: ".boldannounce",
		becomes: "vc_system",
		pretty: "System Messages",
		tooltip: "Messages from your client, always enabled",
		required: true
	}
]
```

You can add any new filters there.
Matches is a CSS selector to check that first <span> tag against.
Becomes is the name of the category it's assigned in vchat (just needs to be unique)
Pretty is the name it gets on the slider toggles in the settings page
Tooltip is the tooltip when you mouse over said slider toggle
Required means it's mandatory on every tab
Admin means it only shows up for admins (purely a convenience feature to keep the list smaller for non-admins, this is not a security feature)
